Added constructors to allow multiple running instances of FFmpeg

evandixon opened the pull request:
When doing batch conversions on multi-core computers, it takes less time to have multiple instances of FFmpeg running concurrently, which prior to this PR, was explicitly disabled.  Now it's only disabled by default.

evandixon committed:
Added constructor variable to allow multiple instances of FFmpeg

evandixon committed:
Excluded packages in .gitignore

evandixon committed:
Added nuspec file for use with build systems

evandixon committed:
Added more constructors

evandixon committed:
Deleted nuspec file in preparation for PR

evandixon committed:
Removed packages folder

evandixon committed:
Revert "Deleted nuspec file in preparation for PR"

This reverts commit a194876f655633c1cf41cda1c6361d473c5a1957.

evandixon committed:
Replaced version with placeholder

danludwig commented:
@evandixon Have you had success with this fork of the library? We are also using a fork of MediaToolkit and running into some concurrency issues currently in tests (since we have multiple tests that use this library running in parallel with xunit). The error we seem to be getting when there is concurrent access is `System.ComponentModel.Win32Exception : Access is denied`.

I am thinking of pulling your changes into our fork as well, but wanted to check with you first to make sure it is working as expected for you, and to check and see if you have made any additional changes that are not in this pull request.

Thanks,

Dan Ludwig

evandixon committed:
Fixed threading issues & improved code quality

- There is no longer a singular FFmpeg process, because outside of the conversion function, it was only used during EngineBase.Dispose, where nothing happened because the conversion function disposes it
- Added thread safety to EnsureFFmpegFileExists
- Changed default path of the ffmpeg executable to be in the temporary directory, to avoid littering the root of the drive.  Included in this path is a new Guid, which will allow it to be cleaned up without affecting other instances of Engine
- Removed the event handler from FFmpegProcess.ErroDataReceived.  In general (I don't know about this particular case), not removing event handlers can prevent the garbage collector from cleaning up the class

evandixon commented:
@danludwig Since I originally made this PR, I've started doing similar things for other tools.  I've just pushed another commit that should fix additional concurrency issues.

My original method of concurrent execution unfortunately overlooked exceptions caused by concurrent execution.  I've since fixed it so that I'm notified when any exceptions are thrown.  My latest commit has been tested in two ways:

- Concurrently executing Engine.Convert with a single instance of Engine
- Concurrently executing Engine.Convert with each execution using a unique instance of Engine
